### PR TITLE
feat(privacy): support hiding polyline start, end and choosing points

### PR DESCRIPTION
**Pull request by EINDEX**
**Little privacy enhancement**

Adding 4 environment variables to support hiding area for polyline start end and chosen points.

Using [this](https://developers.google.com/maps/documentation/utilities/polylineutility) to making `IGNORE_POLYLINE`.

```shell
IGNORE_START_END_RANGE = 200 # ignore meters for each polyline start and end.

IGNORE_RANGE = 200 # ignore meters for each point in the below polyline. 
IGNORE_POLYLINE = ktjrFoemeU~IorGq}DeB # a polyline include point you want to ignore. 

# Do filter before saving to the database, you will lose some data, but you can protect your privacy when you use public repo. enable for set 1, disable for set 0, or unset it.
IGNORE_BEFORE_SAVING = 1
```

### CODE REVIEW

**Comment by yihong0618** [comment]
@EINDEX one thing we need to discuss that should we make it default?

File: scripts/polyline_processor.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    for point in pl:
+        if point_in_list_points_range(
+            point, [start_point, end_point], IGNORE_START_END_RANGE
+        ):
+            continue
+        if point_in_list_points_range(point, IGNORE_POLYLINE, IGNORE_RANGE):
+            continue
+
+        new_pl.append(point)
+    if len(new_pl) == 0:
```
Comment (yihong0618) [inline scripts/polyline_processor.py:43]: if not new_pl:

File: scripts/polyline_processor.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import polyline
+import os
+from haversine import haversine
+
+IGNORE_POLYLINE = (
+    polyline.decode(os.getenv("IGNORE_POLYLINE"))
+    if os.getenv("IGNORE_POLYLINE")
+    else []
+)
+IGNORE_RANGE = int(os.getenv("IGNORE_RANGE", "0")) / 1000
```
Comment (yihong0618) [inline scripts/polyline_processor.py:11]: if IGNORE_RANGE is str it will break I wonder, do we need to info that to users?

**Review by yihong0618 — APPROVED** [review]

**Comment by yihong0618** [comment]
very neat code, thanks

**Comment by EINDEX** [comment]
> @EINDEX one thing we need to discuss that should we make it default?

From the security angle, I suggest set as default. 

But I just have experience with Strava, apple watch(GPX), and Nike data. 
Only Strava has hiding capability.

I think this can be decided by the reviewers' opinion.

File: scripts/polyline_processor.py (hunk 2 above)
Comment (EINDEX) [reply scripts/polyline_processor.py]: Yes changed.

**Comment by ben-29** [comment]
@EINDEX  First of all, well done. However, there is a problem with this implementation, which is that run laps around the playground will be truncated. Is there a way to avoid this?
As far as I know, Strava only hides the front and rear parts, is it possible to consider this approach?
image

**Comment by yihong0618** [comment]
@EINDEX I think we should make it default `false` since we have some users run it, and do not break.
then we add the INFO in README @ben-29 what do you think?

**Comment by yihong0618** [comment]
> @EINDEX First of all, well done. However, there is a problem with this implementation, which is that run laps around the playground will be truncated. Is there a way to avoid this? As far as I know, Strava only hides the front and rear parts, is it possible to consider this approach? image

starva way +1

**Comment by EINDEX** [comment]
@ben-29 Good Idea! will do it soon.

**Comment by EINDEX** [comment]
@yihong0618 @ben-29 add that logic, welcome review.

https://github.com/yihong0618/running_page/pull/441/commits/fc3de6e090202ed4f170c2ad308d9500c02a72e9

**Comment by vlin** [comment]
image
[Diff Link](https://github.com/yihong0618/running_page/commit/fc3de6e090202ed4f170c2ad308d9500c02a72e9#diff-ddaabe903d68d8842158c701253989238155f46b4ee643c6dec30b8ee86ff6cfR36-R42)

@EINDEX 
Here are two small suggestions:
1. Should `range_hidding` be named: `range_hiding`?
2. The data based on polyline may be a bit too much, it would be better if it can be considered to make it more performant.
Please correct me if I'm inconsiderate.🤝
```
def range_hiding(polyline: List[Tuple[float]], points: List[Tuple[float]], distance: int) -> List[Tuple[float]]:
    return [point for point in polyline if not point_in_list_points_range(point, points, distance)]
```

**Review by vlin — COMMENTED** [review]
I don't have any other comments with this PR.

@yihong0618 @ben-29 
The choice to merge this or not is yours to make.

**Comment by yihong0618** [comment]
LGTM

**Comment by yihong0618** [comment]
@EINDEX Thanks for this
@ben-29 @vlin thank you for the review~

**Pull request merged**